ratelimits/names.go

```diff
@@ -0,0 +1,165 @@
+	FailedAuthorizationsPerAccount
+
+	// CertificatesPerDomainPerAccount uses bucket key 'enum:regId:domain',
+	// where name is the a name in a certificate issued to the account matching
+	// regId.
+	CertificatesPerDomainPerAccount
+
+	// CertificatesPerFQDNSetPerAccount uses bucket key 'enum:regId:fqdnSet',
+	// where nameSet is a set of names in a certificate issued to the account
+	// matching regId.
+	CertificatesPerFQDNSetPerAccount
+)
+
+// nameToString is a map of Name values to string names.
+var nameToString = map[Name]string{
+	UsageRequestsPerIPv4Address:      "UsageRequestsPerIPv4Address",
+	UsageRequestsPerIPv6Range:        "UsageRequestsPerIPv6Range",
+	InfoRequestsPerIPv4Address:       "InfoRequestsPerIPv4Address",
+	InfoRequestsPerIPv6Range:         "InfoRequestsPerIPv6Range",
+	NewRegistrationsPerIPv4Address:   "NewRegistrationsPerIPv4Address",
+	NewRegistrationsPerIPv6Range:     "NewRegistrationsPerIPv6Range",
+	NewOrdersPerAccount:              "NewOrdersPerAccount",
+	FailedAuthorizationsPerAccount:   "FailedAuthorizationsPerAccount",
+	CertificatesPerDomainPerAccount:  "CertificatesPerDomainPerAccount",
+	CertificatesPerFQDNSetPerAccount: "CertificatesPerFQDNSetPerAccount",
+}
+
+// validIPv4Address validates that the provided string is a valid IPv4 address.
+func validIPv4Address(id string) error {
+	ip := net.ParseIP(id)
+	if ip == nil || ip.To4() == nil {
+		return fmt.Errorf(
+			"invalid address, %q must be IPv4 address", id)
+	}
+	return nil
+}
+
+// validIPv6RangeCIDR validates that the provided string is formatted is an IPv6
+// CIDR range with a /48 mask.
+func validIPv6RangeCIDR(id string) error {
+	_, ipNet, err := net.ParseCIDR(id)
+	if err != nil {
+		return fmt.Errorf(
+			"invalid CIDR, %q must be an IPv6 CIDR range", id)
+	}
+	ones, _ := ipNet.Mask.Size()
+	if ones != 48 {
+		// This also catches the case where the range is an IPv4 CIDR, since an
+		// IPv4 CIDR can't have a /48 subnet mask - the maximum is /32.
+		return fmt.Errorf(
+			"invalid CIDR, %q must be /48", id)
+	}
+	return nil
+}
+
+// validateRegId validates that the provided string is a valid ACME regId.
+func validateRegId(id string) error {
+	_, err := strconv.ParseUint(id, 10, 64)
+	if err != nil {
+		return fmt.Errorf("invalid regId, %q must be an ACME registration Id", id)
+	}
+	return nil
+}
+
+func validateIdForName(name Name, id string) error {
+	switch name {
+	case UsageRequestsPerIPv4Address, InfoRequestsPerIPv4Address:
+		// 'enum:ipv4address'
+		return validIPv4Address(id)
+
+	case UsageRequestsPerIPv6Range, InfoRequestsPerIPv6Range:
+		// 'enum:ipv6rangeCIDR'
+		return validIPv6RangeCIDR(id)
+
+	case NewOrdersPerAccount:
+		// 'enum:regId'
+		return validateRegId(id)
+
+	case NewRegistrationsPerIPv4Address,
+		NewRegistrationsPerIPv6Range,
+		FailedAuthorizationsPerAccount,
+		CertificatesPerDomainPerAccount,
+		CertificatesPerFQDNSetPerAccount:
+		return fmt.Errorf("overrides are not supported for limit %q", name)
```

Comment:
https://letsencrypt.org/docs/too-many-registrations-for-this-ip
https://letsencrypt.org/docs/failed-validation-limit
https://letsencrypt.org/docs/duplicate-certificate-limit